source/extensions/http/custom_response/redirect_policy/redirect_policy.cc

```diff
@@ -126,18 +159,26 @@ ::Envoy::Http::FilterHeadersStatus RedirectPolicy::encodeHeaders(
         }
       });
 
+  ::Envoy::Http::Utility::Url absolute_url;
+  std::string uri(uri_ ? *uri_
+                       : ::Envoy::Http::Utility::newPath(*redirect_action_, *downstream_headers));
+  if (!absolute_url.initialize(uri, false)) {
+    stats_.custom_response_invalid_uri_.inc();
+    ENVOY_BUG(!static_cast<bool>(uri_),
+              "uri should not be invalid as this was already validated during config load");
+    return ::Envoy::Http::FilterHeadersStatus::Continue;
+  }
   downstream_headers->setScheme(absolute_url.scheme());
   downstream_headers->setHost(absolute_url.hostAndPort());
 
-  auto path_and_query = path_;
+  auto path_and_query = absolute_url.pathAndQueryParams();
   // Strip the #fragment from Location URI if it is present. Envoy treats
   // internal redirect as a new request and will reject it if the URI path
   // contains #fragment.
-  const auto fragment_pos = path_.find('#');
-  path_and_query = path_.substr(0, fragment_pos);
-
-  downstream_headers->setPath(path_);
+  const auto fragment_pos = path_and_query.find('#');
```

Comment:
Will it make sense to reject '#' in the config parsing step?

Reply:
Great point! Yes, it does make sense and we need both, since `#` could be in the original request and allow changing only host and not path for the redirect.